packages/next/src/middleware/sitemaps/createSitemaps.ts

```diff
@@ -0,0 +1,279 @@
+import { XMLParser } from 'fast-xml-parser';
+import { NextRequest } from 'next/server.js';
+import {
+  HandleSitemapRequestsConfig,
+  FAUST_PAGES_PATHNAME,
+} from './handleSitemapRequests.js';
+import {
+  createSitemap,
+  createSitemapIndex,
+  SitemapSchemaSitemapElement,
+  SitemapSchemaUrlElement,
+  isArray,
+  trimSlashes,
+  isUndefined,
+} from './sitemapUtils.js';
+
+/**
+ * TypeScript representation of a parsed XML sitemap
+ */
+export interface ParsedSitemap {
+  urlset: {
+    url: SitemapSchemaUrlElement[];
+  };
+}
+
+/**
+ * TypeScript representation of a parsed XML sitemap index
+ */
+export interface ParsedSitemapIndex {
+  sitemapindex: {
+    sitemap: SitemapSchemaSitemapElement[];
+  };
+}
+
+/**
+ * Creates the root XML sitemap index (e.g. /sitemap.xml) that lists all the
+ * sitemaps provided as the sitemapPaths property in the config, in addition to
+ * a sitemap for the Next.js pages provided as the pages property in the config.
+ *
+ * @param {NextRequest} req The Next.js middleware request object
+ * @param {HandleSitemapRequestsConfig} normalizedConfig A normalized config object
+ * @returns {Response|undefined}
+ */
+export async function createRootSitemapIndex(
+  req: NextRequest,
+  normalizedConfig: HandleSitemapRequestsConfig,
+): Promise<Response | undefined> {
+  const { pages, sitemapPathsToIgnore, replaceUrls, wpUrl } = normalizedConfig;
+  const { pathname, origin } = new URL(req.url);
+  let sitemaps: SitemapSchemaSitemapElement[] = [];
+
+  if (!isUndefined(pages) && isArray(pages) && pages.length) {
+    sitemaps = [
+      ...sitemaps,
+      { loc: `${trimSlashes(origin)}/${trimSlashes(FAUST_PAGES_PATHNAME)}` },
+    ];
+  }
+
+  const wpSitemapUrl = `${trimSlashes(wpUrl)}/${trimSlashes(pathname)}`;
+  const res = await fetch(wpSitemapUrl);
```

Comment:
Will that request could be used to DDoS the Site? If for example a bot trolls the /sitemap.xml for example

Reply:
Yes, I suppose this could be true. This could be true of any endpoint that does a fetch request though as well. We could maybe build some internal cache that will store this requested data on a x minute interval. This could also be configured by the user.

For example: https://github.com/blakewilson/next-wp-template-hierarchy/blob/main/lib/seedQueryCache.mjs